### PR TITLE
Add performance tests for localized site

### PR DESCRIPTION
**Pull request by swissspidy**
* Installs `de_DE` language pack to ensure test site can be localized
* Test both TT1 and TT3 with `de_DE` active
* Adds memory usage to Server-Timing to directly see impact on memory
* Adds admin tests to see impact on the backend
	* Does not measure LCP as it seems useless there
* Update `log-results.js` to ensure all new tests are sent to the Code Vitals dashboard too
* Fixes Playwright reporter to improve error reporting on CI.

Example output on CI:

Screenshot 2023-11-06 at 21 12 17
Screenshot 2023-11-06 at 21 12 07

Trac ticket: https://core.trac.wordpress.org/ticket/59815

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: .github/workflows/performance.yml [hunk 1]
```diff
@@ -162,6 +162,14 @@ jobs:
         run: |
           npm run env:cli -- rewrite structure '/%year%/%monthnum%/%postname%/' --path=/var/www/${{ env.LOCAL_DIR }}
 
+      - name: Install additional languages
+        run: |
+          npm run env:cli -- language core install de_DE --path=/var/www/${{ env.LOCAL_DIR }}
+          npm run env:cli -- language plugin install de_DE --all --path=/var/www/${{ env.LOCAL_DIR }}
+          npm run env:cli -- language plugin update de_DE --all --path=/var/www/${{ env.LOCAL_DIR }}
+          npm run env:cli -- language theme install de_DE --all --path=/var/www/${{ env.LOCAL_DIR }}
+          npm run env:cli -- language theme update de_DE --all --path=/var/www/${{ env.LOCAL_DIR }}
```
Comment (felixarntz) [inline .github/workflows/performance.yml]: Why do the plugin and theme ones need to run `update` immediately after `install`? I'm not familiar and it looks a bit weird (also because it's not the case for core), so it would be good to add a comment to explain.

File: tests/performance/playwright.config.js [hunk 2]
```diff
@@ -19,9 +19,7 @@ process.env.TEST_RUNS ??= '20';
 const config = defineConfig( {
 	...baseConfig,
 	globalSetup: require.resolve( './config/global-setup.js' ),
-	reporter: process.env.CI
-		? './config/performance-reporter.js'
-		: [ [ 'list' ], [ './config/performance-reporter.js' ] ],
+	reporter: [ [ 'list' ], [ './config/performance-reporter.js' ] ],
```
Comment (felixarntz) [inline tests/performance/playwright.config.js:22]: What's the rationale for this change?

File: tests/performance/results.js [hunk 3]
```diff
@@ -9,7 +9,9 @@ const { median, getResultsFilename } = require( './utils' );
 
 const testSuites = [
     'home-classic-theme',
+    'home-classic-theme-l10n',
     'home-block-theme',
+    'home-block-theme-l10n',
 ];
```
Comment (felixarntz) [inline tests/performance/results.js]: Why are the `admin` tests not included here?

File: tests/performance/specs/admin-l10n.test.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { camelCaseDashes } from '../utils';
+
+const results = {
+	timeToFirstByte: [],
+	largestContentfulPaint: [],
+	lcpMinusTtfb: [],
```
Comment (felixarntz) [inline tests/performance/specs/admin-l10n.test.js]: Since these aren't populated here, maybe remove?

File: tests/performance/specs/admin.test.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { camelCaseDashes } from '../utils';
+
+const results = {
+	timeToFirstByte: [],
+	largestContentfulPaint: [],
+	lcpMinusTtfb: [],
```
Comment (felixarntz) [inline tests/performance/specs/admin.test.js]: See above.

File: tests/performance/wp-content/mu-plugins/server-timing.php [hunk 6]
```diff
@@ -25,6 +25,8 @@ static function () use ( $server_timing_values, $template_start ) {
 
 				$server_timing_values['total'] = $server_timing_values['before-template'] + $server_timing_values['template'];
 
+				$server_timing_values['memory-usage'] = memory_get_usage();
```
Comment (felixarntz) [inline tests/performance/wp-content/mu-plugins/server-timing.php:33]: Strictly speaking, Server-Timing is not intended for this metric since it's not timing-based. But I guess it works since it's just another number? Maybe worth adding a comment to clarify regardless.

File: tests/performance/specs/admin-l10n.test.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+		} );
+	} );
+
+	const iterations = Number( process.env.TEST_RUNS );
+	for ( let i = 1; i <= iterations; i++ ) {
+		test( `Measure load time metrics (${ i } of ${ iterations })`, async ( {
+			page,
+			metrics,
+		} ) => {
+			await page.goto( '/' );
```
Comment (felixarntz) [inline tests/performance/specs/admin-l10n.test.js]: Is this right? I would expect this to be something like `/wp-admin/index.php`. Or where else do you ensure it's actually visiting the admin URL?

File: tests/performance/specs/admin.test.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		} );
+	} );
+
+	const iterations = Number( process.env.TEST_RUNS );
+	for ( let i = 1; i <= iterations; i++ ) {
+		test( `Measure load time metrics (${ i } of ${ iterations })`, async ( {
+			page,
+			metrics,
+		} ) => {
+			await page.goto( '/' );
```
Comment (felixarntz) [inline tests/performance/specs/admin.test.js]: See above.

**Review by felixarntz — COMMENTED** [review]
Thanks @swissspidy, I like where this is going. A bit of feedback (mostly questions) below.

File: tests/performance/results.js (hunk 3 above)
Comment (swissspidy) [reply tests/performance/results.js]: Because we have too many places where the test suite names are hardcoded, ugh. Fixing now.

File: tests/performance/wp-content/mu-plugins/server-timing.php (hunk 6 above)
Comment (swissspidy) [reply tests/performance/wp-content/mu-plugins/server-timing.php:33]: It's a bit of a gray area, but it's totally valid and values like `Server-Timing: missedCache` or `pageType;desc="index"` (which e.g. Shopify does) are totally supported. This is nice because `Server-Timing` is only way to pass something from the server to the frontend where it can be read by JavaScript using the `PerformanceTiming` API. So it's a nice little trick :-)

Happy to add a comment if it helps 👍

File: tests/performance/playwright.config.js (hunk 2 above)
Comment (swissspidy) [reply tests/performance/playwright.config.js:22]: It's an oversight from the initial implementation. Without this, any errors normally thrown by Playwright are simply swallowed. There is no good reason to do this.

File: tests/performance/results.js [hunk 9]
```diff
@@ -8,8 +8,12 @@ const { join } = require( 'node:path' );
 const { median, getResultsFilename } = require( './utils' );
 
 const testSuites = [
+	'admin',
+	'admin-l10n',
```
Comment (joemcgill) [inline tests/performance/results.js]: Looks like some tabs vs. spaces discrepancy here

File: tests/performance/specs/admin-l10n.test.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+ */
+import { camelCaseDashes } from '../utils';
+
+const results = {
+	timeToFirstByte: [],
+};
+
+test.describe( 'Admin (L10N)', () => {
+	test.beforeAll( async ( { requestUtils } ) => {
+		await requestUtils.activateTheme( 'twentytwentyone' );
```
Comment (joemcgill) [inline tests/performance/specs/admin-l10n.test.js:17]: To what extent does activating a specific theme matter for these two admin tests vs. using the default theme? Are you mainly just trying to maintain consistency from version to version?

File: .github/workflows/performance.yml [hunk 11]
```diff
@@ -162,6 +162,12 @@ jobs:
         run: |
           npm run env:cli -- rewrite structure '/%year%/%monthnum%/%postname%/' --path=/var/www/${{ env.LOCAL_DIR }}
 
+      - name: Install additional languages
+        run: |
+          npm run env:cli -- language core install de_DE --path=/var/www/${{ env.LOCAL_DIR }}
+          npm run env:cli -- language plugin install de_DE --all --path=/var/www/${{ env.LOCAL_DIR }}
+          npm run env:cli -- language theme install de_DE --all --path=/var/www/${{ env.LOCAL_DIR }}
```
Comment (joemcgill) [inline .github/workflows/performance.yml:169]: Any particular reason behind loading `de_DE` over other translations? Does it have any particular attributes or just a random decision?

File: tests/performance/log-results.js [hunk 12]
```diff
@@ -11,8 +11,12 @@ const { median } = require( './utils' );
 
 // The list of test suites to log.
 const testSuites = [
+	'admin',
```
Comment (joemcgill) [inline tests/performance/log-results.js:14]: Do you know if we need to add these to the [codevitals dashboard](https://www.codevitals.run/project/wordpress)?

**Review by joemcgill — APPROVED** [review]
I've got a couple of questions, but overall this looks good to me.

File: tests/performance/specs/admin-l10n.test.js (hunk 10 above)
Comment (swissspidy) [reply tests/performance/specs/admin-l10n.test.js:17]: Yes, mainly for consistency

File: .github/workflows/performance.yml (hunk 11 above)
Comment (swissspidy) [reply .github/workflows/performance.yml:169]: Random :)

In the past I tried to see whether a language like Japanese would cause different results but that wasn't the case.

File: tests/performance/log-results.js (hunk 12 above)
Comment (swissspidy) [reply tests/performance/log-results.js:14]: Depends on whether we want to add them there or not (I'd say yes, otherwise what's the point). We would need to ask Riad to update the dashboard.

File: tests/performance/results.js (hunk 9 above)
Comment (swissspidy) [reply tests/performance/results.js]: Well spotted! Too bad we don't have Prettier or ESLint 🙃  Fixing now...

File: tests/performance/log-results.js (hunk 12 above)
Comment (joemcgill) [reply tests/performance/log-results.js:14]: Right, I was thinking the same. Let's make sure to reach out after these get added.

File: tests/performance/compare-results.js [hunk 13]
```diff
@@ -128,6 +135,22 @@ console.log( 'Performance Test Results\n' );
 
 console.log( 'Note: Due to the nature of how GitHub Actions work, some variance in the results is expected.\n' );
 
+/**
+ * Nicely format a given metric's value.
+ * @param {string} metric Metric.
```
Comment (mukeshpanchal27) [inline tests/performance/compare-results.js]:
```suggestion
 * Nicely format a given metric's value.
 *
 * @param {string} metric Metric.
```

File: tests/performance/specs/admin-l10n.test.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+	for ( let i = 1; i <= iterations; i++ ) {
+		test( `Measure load time metrics (${ i } of ${ iterations })`, async ( {
+			admin,
+			metrics,
+		} ) => {
+			await admin.visitAdminPage( '/' );
+
+			const serverTiming = await metrics.getServerTiming();
+
+			for (const [key, value] of Object.entries( serverTiming ) ) {
```
Comment (mukeshpanchal27) [inline tests/performance/specs/admin-l10n.test.js]:
```suggestion
			for ( const [ key, value ] of Object.entries( serverTiming ) ) {
```

File: tests/performance/specs/admin.test.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+	for ( let i = 1; i <= iterations; i++ ) {
+		test( `Measure load time metrics (${ i } of ${ iterations })`, async ( {
+			admin,
+			metrics,
+		} ) => {
+			await admin.visitAdminPage( '/' );
+
+			const serverTiming = await metrics.getServerTiming();
+
+			for (const [key, value] of Object.entries( serverTiming ) ) {
```
Comment (mukeshpanchal27) [inline tests/performance/specs/admin.test.js]:
```suggestion
			for ( const [ key, value ] of Object.entries( serverTiming ) ) {
```

File: tests/performance/specs/home-block-theme-l10n.test.js [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+	for ( let i = 1; i <= iterations; i++ ) {
+		test( `Measure load time metrics (${ i } of ${ iterations })`, async ( {
+			page,
+			metrics,
+		} ) => {
+			await page.goto( '/' );
+
+			const serverTiming = await metrics.getServerTiming();
+
+			for ( const [key, value] of Object.entries( serverTiming ) ) {
```
Comment (mukeshpanchal27) [inline tests/performance/specs/home-block-theme-l10n.test.js]:
```suggestion
			for ( const [ key, value ] of Object.entries( serverTiming ) ) {
```

File: tests/performance/specs/home-classic-theme-l10n.test.js [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+	for ( let i = 1; i <= iterations; i++ ) {
+		test( `Measure load time metrics (${ i } of ${ iterations })`, async ( {
+			page,
+			metrics,
+		} ) => {
+			await page.goto( '/' );
+
+			const serverTiming = await metrics.getServerTiming();
+
+			for (const [key, value] of Object.entries( serverTiming ) ) {
```
Comment (mukeshpanchal27) [inline tests/performance/specs/home-classic-theme-l10n.test.js]:
```suggestion
			for ( const [ key, value ] of Object.entries( serverTiming ) ) {
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @swissspidy for the PR. Overall look solid to me left some nit-pick.

File: tests/performance/specs/admin.test.js (hunk 15 above)
Comment (swissspidy) [reply tests/performance/specs/admin.test.js]: yeah unfortunately we don't have eslint/prettier. this is an issue in all files.

**Review by felixarntz — APPROVED** [review]
@swissspidy Thanks for the updates, LGTM!

Are these values just going to appear in the codevitals dashboard, or do we have to update that project as well?

**Comment by swissspidy** [comment]
I will ping Riad to update the dashboard once committed

File: tests/performance/wp-content/mu-plugins/server-timing.php [hunk 18]
```diff
@@ -25,6 +25,11 @@ static function () use ( $server_timing_values, $template_start ) {
 
 				$server_timing_values['total'] = $server_timing_values['before-template'] + $server_timing_values['template'];
 
+				// While values passed via Server-Timing are intended to be durations,
+				// any numeric value can actually be passed.
+				// This is a nice little trick as it allows to easily get this information in JS.
```
Comment (mukeshpanchal27) [inline tests/performance/wp-content/mu-plugins/server-timing.php]:
```suggestion
				/*
				 * While values passed via Server-Timing are intended to be durations,
				 * any numeric value can actually be passed.
				 * This is a nice little trick as it allows to easily get this information in JS.
				 */
```

Use multiline comment.

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks @swissspidy for the update. LGTM!

File: tests/performance/wp-content/mu-plugins/server-timing.php (hunk 18 above)
Comment (Mamaduka) [reply tests/performance/wp-content/mu-plugins/server-timing.php]: ~~JS doesn't follow the exact same coding standards as PHP; both conventions are acceptable here.~~

Oops. Wrong file 😅

Comment (swissspidy) [reply tests/performance/wp-content/mu-plugins/server-timing.php]: @Mamaduka Even in JS it's not really _accepted_ AFAIK, see https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/#multi-line-comments and https://make.wordpress.org/core/2022/11/24/proposal-amend-the-inline-documentation-standards-for-multi-line-comments/. There was another proposal 5 years ago at https://make.wordpress.org/core/2018/12/18/proposed-revisions-to-javascript-coding-standards/ but it didn't result in any changes to the handbook. That said, it's still common in Gutenberg 🤷

**Comment by swissspidy** [comment]
Committed in https://core.trac.wordpress.org/changeset/57083
